File: syncloud/src/main/java/org/syncloud/android/tasks/AsyncResult.java
package org.syncloud.android.tasks;

import com.google.common.base.Optional;

public class AsyncResult<T> {
    private Optional<T> value = Optional.absent();
    private Optional<Throwable> exception = Optional.absent();

    public AsyncResult(Optional<T> value, Optional<Throwable> exception) {
        this.value = value;
        this.exception = exception;
    }

    static public <T> AsyncResult<T> exception(Throwable exception) {
        return new AsyncResult<T>(Optional.<T>absent(), Optional.of(exception));
    }

    static public <T> AsyncResult<T> value(T value) {
        return new AsyncResult<T>(Optional.of(value), Optional.<Throwable>absent());
    }

    public boolean hasValue() {
        return value.isPresent();
    }

    public T getValue() {
        return value.get();
    }
}

File: syncloud/src/main/java/org/syncloud/android/tasks/AsyncResult.kt
package org.syncloud.android.tasks

import com.google.common.base.Optional
import org.syncloud.android.tasks.AsyncResult

class AsyncResult<T>(private val value: Optional<T>, val exception: Optional<Throwable>) {
    fun hasValue(): Boolean {
        return value.isPresent
    }

    fun getValue(): T {
        return value.get()
    }

    fun getException(): Throwable {
        return exception.get()
    }

    companion object {
        @JvmStatic
        fun <T> exception(exception: Throwable): AsyncResult<T> {
            return AsyncResult(Optional.absent(), Optional.of(exception))
        }

        @JvmStatic
        fun <T> value(value: T): AsyncResult<T> {
            return AsyncResult(Optional.of(value), Optional.absent())
        }
    }
}